cmake: modules: Update zephyr_code_relocate API to support relocating libraries

danieldegrasse opened the pull request:
~Add zephyr_library_relocate, a helper macro intended to allow entire Zephyr libraries to be relocated to a new memory region.~ 

Update `zephyr_code_relocate` macro API to support the following:
- A comma separated list of files (may be used with CMake generator expressions)
- A CMake list of files (useful for `file(GLOB ...)`) directives
- relocating all sources of CMake targets, such as all sources in a Zephyr library

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

carlocaione commented:
Please update docs and samples

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1281,6 +1281,20 @@ function(zephyr_linker_sources location)
   endforeach()
 endfunction(zephyr_linker_sources)
 
+function(zephyr_library_relocate location)
```
instead of creating a new function for this, which needs additional processing in the toplevel CMakeLists.txt file, I would much prefer extending the existing `zephyr_code_relocate()` function
https://github.com/zephyrproject-rtos/zephyr/blob/04684c2837dd8af952aff85885c3be7131c9ae37/cmake/modules/extensions.cmake#L1302

to accept generator expressions.

That avoids having one more function to maintain, and keeps other code clean.

Users can then do something like:
```
zephyr_code_relocate($<TARGET_PROPERTY:library,SOURCES> location)
```

A genex is not configure time processing dependent, making the code itself more robust (generally speaking).

As a small benefit, it will even allow downstream users to relocate files in a Zephyr library without patching Zephyr itself.

danieldegrasse replied:
I've been looking into this, and I'm not sure how we can generically support CMake generator expressions in this function. If there's a CMake feature I'm unaware of you think might enable this, please let me know.

The core issue is that CMake generators don't seem to have a method of iterating over all files in the target's source property, so we can't handle relative files without changes to the python script. Even if these changes are made, I still don't know of a method to convert the file names to absolute paths using genex expressions, so the benefit you mentioned of downstream users being able to relocate a Zephyr library won't exist.

Another potentially more robust option would be to use the [TARGET_OBJECTS](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_OBJECTS) property directly, since this is what the `gen_relocate_app.py` script will do internally. This would require API changes for the `zephyr_code_relocate` macro though, which I'd prefer to avoid.

tejlmand replied:
> The core issue is that CMake generators don't seem to have a method of iterating over all files in the target's source property, so we can't handle relative files without changes to the python script. Even if these changes are made, I still don't know of a method to convert the file names to absolute paths using genex expressions, so the benefit you mentioned of downstream users being able to relocate a Zephyr library won't exist.

Easy.
Add this code to any sample CMakeLists.txt and try to run `ninja kernel_sources`:
```
add_custom_target(kernel_sources COMMAND ${CMAKE_COMMAND} -E echo $<TARGET_PROPERTY:kernel,SOURCE_DIR>/$<JOIN:$<TARGET_PROPERTY:kernel,SOURCES>,$<COMMA>$<TARGET_PROPERTY:kernel,SOURCE_DIR>/> )
```

That will print:
```
/projects/github/ncs/zephyr/kernel/main_weak.c,/projects/github/ncs/zephyr/kernel/banner.c,/projects/github/ncs/zephyr/kernel/device.c,/projects/github/ncs/zephyr/kernel/errno.c,/projects/github/ncs/zephyr/kernel/fatal.c,/projects/github/ncs/zephyr/kernel/init.c,/projects/github/ncs/zephyr/kernel/kheap.c,/projects/github/ncs/zephyr/kernel/mem_slab.c,/projects/github/ncs/zephyr/kernel/thread.c,/projects/github/ncs/zephyr/kernel/version.c,/projects/github/ncs/zephyr/kernel/idle.c,/projects/github/ncs/zephyr/kernel/mailbox.c,/projects/github/ncs/zephyr/kernel/msg_q.c,/projects/github/ncs/zephyr/kernel/mutex.c,/projects/github/ncs/zephyr/kernel/queue.c,/projects/github/ncs/zephyr/kernel/sem.c,/projects/github/ncs/zephyr/kernel/stack.c,/projects/github/ncs/zephyr/kernel/system_work_q.c,/projects/github/ncs/zephyr/kernel/work.c,/projects/github/ncs/zephyr/kernel/sched.c,/projects/github/ncs/zephyr/kernel/condvar.c,/projects/github/ncs/zephyr/kernel/xip.c,/projects/github/ncs/zephyr/kernel/timeout.c,/projects/github/ncs/zephyr/kernel/timer.c,/projects/github/ncs/zephyr/kernel/poll.c,/projects/github/ncs/zephyr/kernel/mempool.c
```

Absolute path to all C files included in the kernel library, separated by a comma.

We probably want to provide an easy to use API for the users for the repeated pattern of including a library.
But if we support any generator expression, then we can easily create an easy to use flag.

For example:
```
zephyr_code_relocate(FILES <list of files or generator expression> LOCATION <location>)
```
but even allow a easy to use argument to the specific library case, which then will do a complex generator expression behind the scenes:
```
zephyr_code_relocate(LIBRARY <lib> LOCATION <location>)
```

> Another potentially more robust option would be to use the [TARGET_OBJECTS](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_OBJECTS) property directly

With generic genex support, anything is possible.
The user of the function decide which expressions to pass to the function.

> This would require API changes for the zephyr_code_relocate macro though, which I'd prefer to avoid.

Fair point, but considering this PR and #49454, then I believe it's time to cleanup the API for this function and ensure it can handle more use-cases.

tejlmand replied:
@danieldegrasse if you're not in a hurry for this, i will be happy to take a look at improving the function myself when having the time, but you are of course more than welcome to play around with this yourself.

danieldegrasse replied:
I've updated the `zephyr_code_relocate()` macro based on your suggestions above. Thanks for the pointers- my knowledge on CMake is a bit lacking (especially regarding generator expressions).

tejlmand reviewed (changes requested):
Please consider if the existing function can be extended to support generator expressions instead.

danieldegrasse committed:
scripts: build: update gen_relocate_app.py to use new argument separators

Update gen_relocate_app.py to use "|" to separate code relocation
directives, and ";" to separate multiple files in a relocation directive.
This will enable multiple files to be passed to zephyr_code_relocate,
as well as multiple files to be passed from a CMake generator expression.
The script will then seperate these files and relocate each according to
the arguments given to zephyr_code_relocate.

Note! This commit will break support for zephyr_code_relocate until
the CMake function is updated

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
cmake: update API for zephyr_code_relocate to support relocating libraries

Update API for zephyr_code_relocate to support cmake generator expressions,
as well as relocating libraries.

zephyr_code_relocate can now accept a target name to the LIBRARY argument,
which will be converted into a set of source files from that
target to relocate.

Alternatively, files can be passed as a space separated list
or CMake generator expression. This allows users more
flexibility when relocating files. Glob matching functionality is still
available, although the preferred method to do this would now be:

file(GLOB relocate_sources "src/*.c")
zephyr_code_relocate(FILES ${relocate_sources} LOCATION <location>)

Note! This commit breaks support for zephyr_code_relocate until in tree
usages of the API are updated to the new format.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
treewide: update usage of zephyr_code_relocate

Update usage of zephyr_code_relocate to follow new API. The old method
of relocating a file was the following directive:

zephyr_code_relocate(file location)

The new API for zephyr_code_relocate uses the following directive:

zephyr_code_relocate(FILES file LOCATION location)

update in tree usage to follow this model. Also, update the NXP HAL SHA,
as NXP's HAL uses this macro as well.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
doc: code-relocation: Update usage of zephyr_code_relocate

Update usage of zephyr_code_relocate to new API, and add examples of
relocating a library target, as well as using multiple files in list or
CMake generator expressions.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

tejlmand commented on doc/kernel/code-relocation.rst:
```diff
@@ -76,21 +76,30 @@ This section shows additional configuration options that can be set in
+
+     file(GLOB sources "file*.c")
+     zephyr_code_relocate(FILES ${sources} LOCATION SRAM)
+     zephyr_code_relocate(FILES $<TARGET_PROPERTY:my_tgt,SOURCES> LOCATION SRAM)
```
as described in other comment, then I believe this example will not work unfortunately, because the python relocation script will expect the files to be comma separated,

But let's wait with adjusting this until the final approach has been decided.

tejlmand reviewed (changes requested):
Thanks for doing this.

Great improvement, but a couples of details we need to consider / discuss before final approval.

danieldegrasse commented on cmake/modules/extensions.cmake:
```diff
@@ -1283,13 +1283,67 @@ endfunction(zephyr_linker_sources)
+      endforeach()
+      set(file_list "$<JOIN:${file_names},$<COMMA>>")
+    else()
+      # Generator expression is present in file list. Leave the list untouched.
+      set(file_list ${CODE_REL_FILES})
```
I think the substitution method is probably best, with some documentation provided to the user. It is unfortunate, but if we don't perform some type of modification on the user input then the other alternative I can see be to change the genex directive in [target_relocation.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/ae9de202220ce51e1ed61157f73d3800f84c6b06/cmake/linker/ld/target_relocation.cmake#L21) to something like the following:

```
-i \"$<JOIN:$<TARGET_PROPERTY:code_data_relocation_target,COMPILE_DEFINITIONS>,$<COMMA>>\"
```
which would require the genex passed to `zephyr_library_relocate` to be modified, otherwise this `JOIN` directive would change it as well. So either way, we are left needing to change user input.

If we wanted to place the responsibility on the caller, we could update the `gen_relocate_app.py` script to print a warning when it detects a relocation directive without `:` present, which would be the failure mode of the script when a genex expression was passed without `,` subsituted

danieldegrasse replied:
@tejlmand What are your thoughts here? I'd like to move this PR forwards, as #50792 depends on a feature like this being available. Once we settle on a method here I can make other requested updates

danieldegrasse committed:
tests: code_relocation: Add additional test cases for library and genex

Add test cases for generator expressions and library relocation to
the code_relocation test, in order to verify functionality for these
new API features.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@tejlmand I've taken another look at this PR and updated `zephyr_code_relocate` to accept CMake generator expressions directly. I've also extended the `tests/application_development/code_relocation` to test relocating a library, as well as relocating files using a generator expression. Welcome to any feedback you have for these changes!

danieldegrasse commented on doc/kernel/code-relocation.rst:
```diff
@@ -76,21 +76,30 @@ This section shows additional configuration options that can be set in
 * Multiple regions can also be appended together such as: SRAM2_DATA_BSS.
   This will place data and bss inside SRAM2.
 
+* Multiple files can be passed to the FILES argument, or CMake generator
+  expressions can be used to relocate a comma-separated list of files
```
I've removed this note in the latest push, so that regular expressions are no longer documented.

danieldegrasse commented on tests/application_development/code_relocation/CMakeLists.txt:
```diff
@@ -9,18 +9,19 @@ FILE(GLOB app_sources src/*.c)
 
 if (CONFIG_RELOCATE_TO_ITCM)
-zephyr_code_relocate(../../../lib/libc/minimal/source/string/string.c ITCM_TEXT)
+zephyr_code_relocate(FILES ../../../lib/libc/minimal/source/string/string.c
```
Fixed in latest push

danieldegrasse commented on scripts/build/gen_relocate_app.py:
```diff
@@ -433,16 +433,25 @@ def create_dict_wrt_mem():
+            if not glob_results:
+                warnings.warn("File: "+file_glob+" Not found")
+                continue
+            file_name_list.extend(glob_results)
```
I've added a deprecation warning to this script, directing the user to use `file(GLOB)` over regex

danieldegrasse commented on tests/application_development/code_relocation/test_lib/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 NXP
+
+add_library(test_lib STATIC "")
```
I'm not particularly happy with how this library is declared- the goal is to add a library that has the same properties as an in tree Zephyr library, without needing to add a "dummy" library somewhere outside of the sample. I could also use the same approach Zephyr uses for external drivers to add these files, but felt that was unwieldy for a simple test.

danieldegrasse commented on tests/application_development/code_relocation/test_lib/test_lib1.c:
```diff
@@ -0,0 +1,43 @@
+#include <zephyr/sys/printk.h>
+#include <zephyr/ztest.h>
+
+__in_section(data, sram2, var) uint32_t var_lib1_sram2_data = 10U;
```
@tari When testing these changes, I noticed that these directives can mask errors in the test (on Cortex-M). Since this directive moves symbols to `data.sram2.var` unconditionally, these variables will still be relocated if `zephyr_code_relocate` is not called on this file, so long as another file in the build is relocated to `sram2`, since then the section name will be defined.

I'm interested to know what other approaches we could use here (particularly if `volatile` would prevent these variables from being moved to RODATA). The other option I can see is creating new RAM sections for each file (SRAM2, SRAM3, SRAM4, etc), so that if the file isn't relocated the required linker section won't be present.

tari replied:
I don't think `volatile` would be reliable; it would just enforce that reads always actually read without asserting anything about writability. Creating different sections for each file seems messy but reasonable to me: this is a test so it seems okay if it needs to do some weird things in order to verify behavior that's somewhat complex.

carlescufi commented:
@carlocaione can you please take a look?

carlescufi commented:
@tejlmand please re-review

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1283,22 +1285,85 @@ endfunction(zephyr_linker_sources)
     set(copy_flag NOCOPY)
   endif()
+  if(CODE_REL_PHDR)
+    set(CODE_REL_LOCATION "${CODE_REL_LOCATION}\ :${CODE_REL_PHDR}")
```
exactly why is an escaped space needed here ?
That doesn't seem to have been necessary before.

danieldegrasse replied:
The PHDR directive was previously paired with the relocation location, as you can see in this diff: https://github.com/zephyrproject-rtos/zephyr/pull/50791/files#diff-7087c4d7b0a6595b501520d5eb10ed5ecb01b3540831e3a141a6a662632ac940R20. This directive makes sure that the PHDR is still paired with the code location, so it can be parsed by `gen_relocate_app.py`. `gen_relocate_app.py` treats the PHDR as being paired with the relocation location, so this line preserves that behavior with the new API. An alternative here would be to integrate PHDR as a parameter like NOCOPY is used.

tejlmand reviewed (dismissed):
Approved.
A minor comment, but nothing which should prevent merging as it can be updated in followup if need be.
